{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Cellranger count of gene expression data\n",
    "required before guide calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "# adding notebooks to the system path\n",
    "sys.path.insert(0, '/home/southark/notebooks')\n",
    "\n",
    "from guide_calling import *\n",
    "import os\n",
    "\n",
    "import os.path\n",
    "\n",
    "from BarcodeCaller import *\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import seaborn as sns\n",
    "from scipy.signal import argrelextrema\n",
    "from scipy.stats import gaussian_kde\n",
    "from scipy.io import mmread\n",
    "from joblib import Parallel, delayed\n",
    "\n",
    "\n",
    "import scipy.io\n",
    "import gzip\n",
    "import codecs\n",
    "\n",
    "%matplotlib inline  \n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "sns.set_style('white')\n",
    "pd.set_option('display.float_format', lambda x: '%.2f' % x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_and_set_wdir(working_dir):\n",
    "    #create and set working directory\n",
    "    #working_dir = \"/scratch/sequencing/processed_overloading/notebook_test/\"\n",
    "\n",
    "    isdir = os.path.isdir(working_dir)\n",
    "\n",
    "    if isdir:\n",
    "        print('moving to the working directory')\n",
    "        %cd {working_dir}\n",
    "\n",
    "\n",
    "    else:\n",
    "        print('created working directory')\n",
    "        %mkdir {working_dir}\n",
    "        print('moving to the working directory')\n",
    "        %cd {working_dir}      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set experiment cellranger output directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created working directory\n",
      "moving to the working directory\n",
      "/lila/data/norman/southark/rpe1_tfs/23127_RPE1_TFs_CRa\n"
     ]
    }
   ],
   "source": [
    "#set cellranger output dir\n",
    "working_dir = \"/data/norman/southark/rpe1_tfs/23127_RPE1_TFs_CRa\"        \n",
    "create_and_set_wdir(working_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create guide barcode csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/southark/.conda/envs/analysis37/lib/python3.7/site-packages/ipykernel_launcher.py:3: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "/home/southark/.conda/envs/analysis37/lib/python3.7/site-packages/ipykernel_launcher.py:7: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.\n",
      "  import sys\n"
     ]
    }
   ],
   "source": [
    "## create guide barcode csv\n",
    "#import list of expected guides\n",
    "guides = pd.read_csv('/data/norman/southark/guide_libraries/CRISPRa_guide_selection/final_selected/final_combined_paralogs_morf_CRISPRa_library_with_adapters.csv', skipfooter=3)\n",
    "#guides = guides.rename(columns = {'protospacer': 'barcode'})\n",
    "\n",
    "\n",
    "ctrls = pd.read_csv('/data/norman/southark/guide_libraries/CRISPRa_guide_selection/final_selected/Weissman_v3_CRISPRa_filtered_controls_NHDF-Ad_05172023.csv', skipfooter=3)\n",
    "ctrls = ctrls.rename(columns = {'protospacer': 'barcode', 'Unnamed: 0': 'id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>target_gene_id</th>\n",
       "      <th>target_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF_GGAAGCGCGCAGAAGGTTGA</td>\n",
       "      <td>AATF</td>\n",
       "      <td>GGAAGCGCGCAGAAGGTTGA</td>\n",
       "      <td>AATF</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF_GCGTGCGAGTGCGCGGGAAG</td>\n",
       "      <td>AATF</td>\n",
       "      <td>GCGTGCGAGTGCGCGGGAAG</td>\n",
       "      <td>AATF</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF_GCGCAGAAGGTTGAAGGGAT</td>\n",
       "      <td>AATF</td>\n",
       "      <td>GCGCAGAAGGTTGAAGGGAT</td>\n",
       "      <td>AATF</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF_GGGCGTTGCTAGCATGAAGG</td>\n",
       "      <td>AATF</td>\n",
       "      <td>GGGCGTTGCTAGCATGAAGG</td>\n",
       "      <td>AATF</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF_GTGAAGGGATTGGAGCCGTA</td>\n",
       "      <td>AATF</td>\n",
       "      <td>GTGAAGGGATTGGAGCCGTA</td>\n",
       "      <td>AATF</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>non_targeting_GAACGGGCCGTGATCGGACC</td>\n",
       "      <td>non_targeting</td>\n",
       "      <td>GAACGGGCCGTGATCGGACC</td>\n",
       "      <td>Non-Targeting</td>\n",
       "      <td>Non-Targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>non_targeting_GGCCCTCCCCACCGGCACGA</td>\n",
       "      <td>non_targeting</td>\n",
       "      <td>GGCCCTCCCCACCGGCACGA</td>\n",
       "      <td>Non-Targeting</td>\n",
       "      <td>Non-Targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>non_targeting_GTCGTTATCTCGCTATTTCG</td>\n",
       "      <td>non_targeting</td>\n",
       "      <td>GTCGTTATCTCGCTATTTCG</td>\n",
       "      <td>Non-Targeting</td>\n",
       "      <td>Non-Targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>non_targeting_GTGGTTATACCCGACTAGAC</td>\n",
       "      <td>non_targeting</td>\n",
       "      <td>GTGGTTATACCCGACTAGAC</td>\n",
       "      <td>Non-Targeting</td>\n",
       "      <td>Non-Targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>non_targeting_GTCCACAAGACGTGCTCGCA</td>\n",
       "      <td>non_targeting</td>\n",
       "      <td>GTCCACAAGACGTGCTCGCA</td>\n",
       "      <td>Non-Targeting</td>\n",
       "      <td>Non-Targeting</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10979 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    id           name              sequence  \\\n",
       "0            AATF_GGAAGCGCGCAGAAGGTTGA           AATF  GGAAGCGCGCAGAAGGTTGA   \n",
       "1            AATF_GCGTGCGAGTGCGCGGGAAG           AATF  GCGTGCGAGTGCGCGGGAAG   \n",
       "2            AATF_GCGCAGAAGGTTGAAGGGAT           AATF  GCGCAGAAGGTTGAAGGGAT   \n",
       "3            AATF_GGGCGTTGCTAGCATGAAGG           AATF  GGGCGTTGCTAGCATGAAGG   \n",
       "4            AATF_GTGAAGGGATTGGAGCCGTA           AATF  GTGAAGGGATTGGAGCCGTA   \n",
       "..                                 ...            ...                   ...   \n",
       "73  non_targeting_GAACGGGCCGTGATCGGACC  non_targeting  GAACGGGCCGTGATCGGACC   \n",
       "74  non_targeting_GGCCCTCCCCACCGGCACGA  non_targeting  GGCCCTCCCCACCGGCACGA   \n",
       "75  non_targeting_GTCGTTATCTCGCTATTTCG  non_targeting  GTCGTTATCTCGCTATTTCG   \n",
       "76  non_targeting_GTGGTTATACCCGACTAGAC  non_targeting  GTGGTTATACCCGACTAGAC   \n",
       "77  non_targeting_GTCCACAAGACGTGCTCGCA  non_targeting  GTCCACAAGACGTGCTCGCA   \n",
       "\n",
       "   target_gene_id target_gene_name  \n",
       "0            AATF             AATF  \n",
       "1            AATF             AATF  \n",
       "2            AATF             AATF  \n",
       "3            AATF             AATF  \n",
       "4            AATF             AATF  \n",
       "..            ...              ...  \n",
       "73  Non-Targeting    Non-Targeting  \n",
       "74  Non-Targeting    Non-Targeting  \n",
       "75  Non-Targeting    Non-Targeting  \n",
       "76  Non-Targeting    Non-Targeting  \n",
       "77  Non-Targeting    Non-Targeting  \n",
       "\n",
       "[10979 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guides = guides.rename(columns = {'protospacer': 'sequence',\n",
    "                                 # 'sgID':'id',\n",
    "                                 'gene': 'name'})\n",
    "\n",
    "\n",
    "ctrls = ctrls.rename(columns = {'barcode': 'sequence',\n",
    "                                 # 'sgID':'id',\n",
    "                                 'gene': 'name'})\n",
    "\n",
    "guides['sequence'] = guides.sequence.str.upper()\n",
    "#guides['id'] = guides['id'].str.split(',').str[0]\n",
    "\n",
    "\n",
    "guides.drop(columns =['full_sequence'], inplace = True)\n",
    "ctrls.drop(columns =['transcript', 'sgID','source', 'original_rank','entrezgene', 'n_offtargets', 'n_exact_matches',\n",
    "                    'state', 'n_off_strict'], inplace = True)\n",
    "\n",
    "guides = pd.concat([guides, ctrls])\n",
    "\n",
    "\n",
    "guides = guides.assign(target_gene_id = np.where(guides.name.str.contains('non_targeting'),\n",
    "                                                 'Non-Targeting',\n",
    "                                                 guides['name']))\n",
    "\n",
    "guides = guides.assign(target_gene_name = np.where(guides.id.str.contains('non_targeting'),\n",
    "                                                 'Non-Targeting',\n",
    "                                                 guides['name']))\n",
    "\n",
    "#guides['target_gene_id'] = guides['target_gene_id'].str.split('_').str[0]\n",
    "\n",
    "guides"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ## Find sequencing files for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "\n",
    "#set file name search strings\n",
    "SAMPLE_GLOB = '*_GEX*'\n",
    "\n",
    "FB_GLOB = '*_guides*'\n",
    "\n",
    "#perform for each project folder (i.e. if more sequencing was ordered)\n",
    "\n",
    "input_fastq_rna_1=\"/data/norman/southark/datasets/Project_15055/\"\n",
    "\n",
    "unique_samples_1 = sorted(set(map(lambda x: x.split('/')[-1].split('_S')[0].split('Sample_')[0], \n",
    "                                glob(os.path.join(input_fastq_rna_1, '**', SAMPLE_GLOB), recursive=True))))\n",
    "\n",
    "unique_samples_fb_1 = sorted(set(map(lambda x: x.split('/')[-1].split('_S')[0].split('Sample_')[0], \n",
    "                                glob(os.path.join(input_fastq_rna_1, '**', FB_GLOB), recursive=True))))\n",
    "\n",
    "\n",
    "input_fastq_rna_2=\"/data/norman/southark/datasets/Project_15055_B/\"\n",
    "\n",
    "unique_samples_2 = sorted(set(map(lambda x: x.split('/')[-1].split('_S')[0].split('Sample_')[0], \n",
    "                                glob(os.path.join(input_fastq_rna_2, '**', SAMPLE_GLOB), recursive=True))))\n",
    "\n",
    "unique_samples_fb_2 = sorted(set(map(lambda x: x.split('/')[-1].split('_S')[0].split('Sample_')[0], \n",
    "                                glob(os.path.join(input_fastq_rna_2, '**', FB_GLOB), recursive=True))))\n",
    "\n",
    "unique_samples_1.remove('')\n",
    "unique_samples_fb_1.remove('')\n",
    "\n",
    "unique_samples_2.remove('')\n",
    "unique_samples_fb_2.remove('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10A_GEX_IGO_15055_11',\n",
       " '10B_GEX_IGO_15055_16',\n",
       " '11A_GEX_IGO_15055_12',\n",
       " '11B_GEX_IGO_15055_17',\n",
       " '12A_GEX_IGO_15055_13',\n",
       " '12B_GEX_IGO_15055_18',\n",
       " '13A_GEX_IGO_15055_14',\n",
       " '13B_GEX_IGO_15055_19',\n",
       " '14A_GEX_IGO_15055_15',\n",
       " '14B_GEX_IGO_15055_20',\n",
       " '2A_GEX_IGO_15055_1',\n",
       " '2B_GEX_IGO_15055_6',\n",
       " '3A_GEX_IGO_15055_2',\n",
       " '3B_GEX_IGO_15055_7',\n",
       " '4A_GEX_IGO_15055_3',\n",
       " '4B_GEX_IGO_15055_8',\n",
       " '5A_GEX_IGO_15055_4',\n",
       " '5B_GEX_IGO_15055_9',\n",
       " '6A_GEX_IGO_15055_5',\n",
       " '6B_GEX_IGO_15055_10']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_samples_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10A_guides_IGO_15055_31',\n",
       " '10B_guides_IGO_15055_36',\n",
       " '11A_guides_IGO_15055_32',\n",
       " '11B_guides_IGO_15055_37',\n",
       " '12A_guides_IGO_15055_33',\n",
       " '12B_guides_IGO_15055_38',\n",
       " '13A_guides_IGO_15055_34',\n",
       " '13B_guides_IGO_15055_39',\n",
       " '14A_guides_IGO_15055_35',\n",
       " '14B_guides_IGO_15055_40',\n",
       " '2A_guides_IGO_15055_21',\n",
       " '2B_guides_IGO_15055_26',\n",
       " '3A_guides_IGO_15055_22',\n",
       " '3B_guides_IGO_15055_27',\n",
       " '4A_guides_IGO_15055_23',\n",
       " '4B_guides_IGO_15055_28',\n",
       " '5A_guides_IGO_15055_24',\n",
       " '5B_guides_IGO_15055_29',\n",
       " '6A_guides_IGO_15055_25',\n",
       " '6B_guides_IGO_15055_30']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_samples_fb_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10A_GEX_IGO_15055_B_11',\n",
       " '10B_GEX_IGO_15055_B_16',\n",
       " '11A_GEX_IGO_15055_B_12',\n",
       " '11B_GEX_IGO_15055_B_17',\n",
       " '12A_GEX_IGO_15055_B_13',\n",
       " '12B_GEX_IGO_15055_B_18',\n",
       " '13A_GEX_IGO_15055_B_14',\n",
       " '13B_GEX_IGO_15055_B_19',\n",
       " '14A_GEX_IGO_15055_B_15',\n",
       " '14B_GEX_IGO_15055_B_20',\n",
       " '2A_GEX_IGO_15055_B_1',\n",
       " '2B_GEX_IGO_15055_B_6',\n",
       " '3A_GEX_IGO_15055_B_2',\n",
       " '3B_GEX_IGO_15055_B_7',\n",
       " '4A_GEX_IGO_15055_B_3',\n",
       " '4B_GEX_IGO_15055_B_8',\n",
       " '5A_GEX_IGO_15055_B_4',\n",
       " '5B_GEX_IGO_15055_B_9',\n",
       " '6A_GEX_IGO_15055_B_5',\n",
       " '6B_GEX_IGO_15055_B_10']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_samples_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10A_guides_IGO_15055_B_31',\n",
       " '10B_guides_IGO_15055_B_36',\n",
       " '11A_guides_IGO_15055_B_32',\n",
       " '11B_guides_IGO_15055_B_37',\n",
       " '12A_guides_IGO_15055_B_33',\n",
       " '12B_guides_IGO_15055_B_38',\n",
       " '13A_guides_IGO_15055_B_34',\n",
       " '13B_guides_IGO_15055_B_39',\n",
       " '14B_guides_IGO_15055_B_40',\n",
       " '2B_guides_IGO_15055_B_26',\n",
       " '3B_guides_IGO_15055_B_27',\n",
       " '4B_guides_IGO_15055_B_28',\n",
       " '5B_guides_IGO_15055_B_29',\n",
       " '6B_guides_IGO_15055_B_30']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_samples_fb_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "def get_directory_paths(path: str, file_names: list) -> dict:\n",
    "    \"\"\"\n",
    "    Returns the combination of parent directory paths that exist for every file name in a directory.\n",
    "\n",
    "    Args:\n",
    "    path (str): The path to the directory.\n",
    "    file_names (list): A list of unique file names.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary with the file names as keys and a list of parent directory paths as values.\n",
    "    \"\"\"\n",
    "    result = {}\n",
    "    for root, dirs, files in os.walk(path):\n",
    "        for file_name in file_names:\n",
    "            if not result.get(file_name):\n",
    "                result[file_name] = []\n",
    "            for f in files:\n",
    "                #print(f)\n",
    "                if file_name+'_' in f:\n",
    "                    result[file_name].append(os.path.dirname(os.path.join(root, f)))\n",
    "                    break\n",
    "                    \n",
    "    return result\n",
    "\n",
    "\n",
    "def format_fastq_df(path, samples, lib_type):\n",
    "    sample_paths = get_directory_paths(path, samples)\n",
    "\n",
    "    fastq_df = pd.DataFrame([{'sample': key, 'fastqs': value} for key, values in sample_paths.items() for value in values])\n",
    "    fastq_df['library_type'] = lib_type\n",
    "    fastq_df['sample_id'] = fastq_df['sample'].str.split('_').str[0]\n",
    "    \n",
    "    return fastq_df\n",
    "\n",
    "\n",
    "\n",
    "gex_1 = format_fastq_df(input_fastq_rna_1, unique_samples_1, 'Gene Expression')\n",
    "gbc_1 = format_fastq_df(input_fastq_rna_1, unique_samples_fb_1, 'CRISPR Guide Capture')\n",
    "\n",
    "gex_2 = format_fastq_df(input_fastq_rna_2, unique_samples_2, 'Gene Expression')\n",
    "gbc_2 = format_fastq_df(input_fastq_rna_2, unique_samples_fb_2, 'CRISPR Guide Capture')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>fastqs</th>\n",
       "      <th>library_type</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10B_GEX_IGO_15055_16</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10B_GEX_IGO_15055_16</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>11A_GEX_IGO_15055_12</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11A_GEX_IGO_15055_12</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>11B_GEX_IGO_15055_17</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>11B_GEX_IGO_15055_17</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12A_GEX_IGO_15055_13</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>12A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>12A_GEX_IGO_15055_13</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>12A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12B_GEX_IGO_15055_18</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>12B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12B_GEX_IGO_15055_18</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>12B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13A_GEX_IGO_15055_14</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>13A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13A_GEX_IGO_15055_14</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>13A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>13B_GEX_IGO_15055_19</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>13B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>13B_GEX_IGO_15055_19</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>13B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>14A_GEX_IGO_15055_15</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>14A_GEX_IGO_15055_15</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>14A_GEX_IGO_15055_15</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>14B_GEX_IGO_15055_20</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14B_GEX_IGO_15055_20</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>14B_GEX_IGO_15055_20</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>14B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2A_GEX_IGO_15055_1</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2A_GEX_IGO_15055_1</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2A_GEX_IGO_15055_1</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2B_GEX_IGO_15055_6</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2B_GEX_IGO_15055_6</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>2B_GEX_IGO_15055_6</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3A_GEX_IGO_15055_2</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>3A_GEX_IGO_15055_2</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>3A_GEX_IGO_15055_2</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3B_GEX_IGO_15055_7</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>3B_GEX_IGO_15055_7</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>4A_GEX_IGO_15055_3</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>4A_GEX_IGO_15055_3</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>4A_GEX_IGO_15055_3</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>4B_GEX_IGO_15055_8</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>4B_GEX_IGO_15055_8</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>5A_GEX_IGO_15055_4</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>5A_GEX_IGO_15055_4</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>5B_GEX_IGO_15055_9</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>5B_GEX_IGO_15055_9</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>5B_GEX_IGO_15055_9</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>6A_GEX_IGO_15055_5</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>6A_GEX_IGO_15055_5</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>6A_GEX_IGO_15055_5</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>6B_GEX_IGO_15055_10</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>6B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>6B_GEX_IGO_15055_10</td>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>6B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample                                             fastqs  \\\n",
       "0   10A_GEX_IGO_15055_11  /data/norman/southark/datasets/Project_15055/F...   \n",
       "1   10A_GEX_IGO_15055_11  /data/norman/southark/datasets/Project_15055/F...   \n",
       "2   10A_GEX_IGO_15055_11  /data/norman/southark/datasets/Project_15055/F...   \n",
       "3   10B_GEX_IGO_15055_16  /data/norman/southark/datasets/Project_15055/F...   \n",
       "4   10B_GEX_IGO_15055_16  /data/norman/southark/datasets/Project_15055/F...   \n",
       "5   11A_GEX_IGO_15055_12  /data/norman/southark/datasets/Project_15055/F...   \n",
       "6   11A_GEX_IGO_15055_12  /data/norman/southark/datasets/Project_15055/F...   \n",
       "7   11B_GEX_IGO_15055_17  /data/norman/southark/datasets/Project_15055/F...   \n",
       "8   11B_GEX_IGO_15055_17  /data/norman/southark/datasets/Project_15055/F...   \n",
       "9   12A_GEX_IGO_15055_13  /data/norman/southark/datasets/Project_15055/F...   \n",
       "10  12A_GEX_IGO_15055_13  /data/norman/southark/datasets/Project_15055/F...   \n",
       "11  12B_GEX_IGO_15055_18  /data/norman/southark/datasets/Project_15055/F...   \n",
       "12  12B_GEX_IGO_15055_18  /data/norman/southark/datasets/Project_15055/F...   \n",
       "13  13A_GEX_IGO_15055_14  /data/norman/southark/datasets/Project_15055/F...   \n",
       "14  13A_GEX_IGO_15055_14  /data/norman/southark/datasets/Project_15055/F...   \n",
       "15  13B_GEX_IGO_15055_19  /data/norman/southark/datasets/Project_15055/F...   \n",
       "16  13B_GEX_IGO_15055_19  /data/norman/southark/datasets/Project_15055/F...   \n",
       "17  14A_GEX_IGO_15055_15  /data/norman/southark/datasets/Project_15055/F...   \n",
       "18  14A_GEX_IGO_15055_15  /data/norman/southark/datasets/Project_15055/F...   \n",
       "19  14A_GEX_IGO_15055_15  /data/norman/southark/datasets/Project_15055/F...   \n",
       "20  14B_GEX_IGO_15055_20  /data/norman/southark/datasets/Project_15055/F...   \n",
       "21  14B_GEX_IGO_15055_20  /data/norman/southark/datasets/Project_15055/F...   \n",
       "22  14B_GEX_IGO_15055_20  /data/norman/southark/datasets/Project_15055/F...   \n",
       "23    2A_GEX_IGO_15055_1  /data/norman/southark/datasets/Project_15055/F...   \n",
       "24    2A_GEX_IGO_15055_1  /data/norman/southark/datasets/Project_15055/F...   \n",
       "25    2A_GEX_IGO_15055_1  /data/norman/southark/datasets/Project_15055/F...   \n",
       "26    2B_GEX_IGO_15055_6  /data/norman/southark/datasets/Project_15055/F...   \n",
       "27    2B_GEX_IGO_15055_6  /data/norman/southark/datasets/Project_15055/F...   \n",
       "28    2B_GEX_IGO_15055_6  /data/norman/southark/datasets/Project_15055/F...   \n",
       "29    3A_GEX_IGO_15055_2  /data/norman/southark/datasets/Project_15055/F...   \n",
       "30    3A_GEX_IGO_15055_2  /data/norman/southark/datasets/Project_15055/F...   \n",
       "31    3A_GEX_IGO_15055_2  /data/norman/southark/datasets/Project_15055/F...   \n",
       "32    3B_GEX_IGO_15055_7  /data/norman/southark/datasets/Project_15055/F...   \n",
       "33    3B_GEX_IGO_15055_7  /data/norman/southark/datasets/Project_15055/F...   \n",
       "34    4A_GEX_IGO_15055_3  /data/norman/southark/datasets/Project_15055/F...   \n",
       "35    4A_GEX_IGO_15055_3  /data/norman/southark/datasets/Project_15055/F...   \n",
       "36    4A_GEX_IGO_15055_3  /data/norman/southark/datasets/Project_15055/F...   \n",
       "37    4B_GEX_IGO_15055_8  /data/norman/southark/datasets/Project_15055/F...   \n",
       "38    4B_GEX_IGO_15055_8  /data/norman/southark/datasets/Project_15055/F...   \n",
       "39    5A_GEX_IGO_15055_4  /data/norman/southark/datasets/Project_15055/F...   \n",
       "40    5A_GEX_IGO_15055_4  /data/norman/southark/datasets/Project_15055/F...   \n",
       "41    5B_GEX_IGO_15055_9  /data/norman/southark/datasets/Project_15055/F...   \n",
       "42    5B_GEX_IGO_15055_9  /data/norman/southark/datasets/Project_15055/F...   \n",
       "43    5B_GEX_IGO_15055_9  /data/norman/southark/datasets/Project_15055/F...   \n",
       "44    6A_GEX_IGO_15055_5  /data/norman/southark/datasets/Project_15055/F...   \n",
       "45    6A_GEX_IGO_15055_5  /data/norman/southark/datasets/Project_15055/F...   \n",
       "46    6A_GEX_IGO_15055_5  /data/norman/southark/datasets/Project_15055/F...   \n",
       "47   6B_GEX_IGO_15055_10  /data/norman/southark/datasets/Project_15055/F...   \n",
       "48   6B_GEX_IGO_15055_10  /data/norman/southark/datasets/Project_15055/F...   \n",
       "\n",
       "       library_type sample_id  \n",
       "0   Gene Expression       10A  \n",
       "1   Gene Expression       10A  \n",
       "2   Gene Expression       10A  \n",
       "3   Gene Expression       10B  \n",
       "4   Gene Expression       10B  \n",
       "5   Gene Expression       11A  \n",
       "6   Gene Expression       11A  \n",
       "7   Gene Expression       11B  \n",
       "8   Gene Expression       11B  \n",
       "9   Gene Expression       12A  \n",
       "10  Gene Expression       12A  \n",
       "11  Gene Expression       12B  \n",
       "12  Gene Expression       12B  \n",
       "13  Gene Expression       13A  \n",
       "14  Gene Expression       13A  \n",
       "15  Gene Expression       13B  \n",
       "16  Gene Expression       13B  \n",
       "17  Gene Expression       14A  \n",
       "18  Gene Expression       14A  \n",
       "19  Gene Expression       14A  \n",
       "20  Gene Expression       14B  \n",
       "21  Gene Expression       14B  \n",
       "22  Gene Expression       14B  \n",
       "23  Gene Expression        2A  \n",
       "24  Gene Expression        2A  \n",
       "25  Gene Expression        2A  \n",
       "26  Gene Expression        2B  \n",
       "27  Gene Expression        2B  \n",
       "28  Gene Expression        2B  \n",
       "29  Gene Expression        3A  \n",
       "30  Gene Expression        3A  \n",
       "31  Gene Expression        3A  \n",
       "32  Gene Expression        3B  \n",
       "33  Gene Expression        3B  \n",
       "34  Gene Expression        4A  \n",
       "35  Gene Expression        4A  \n",
       "36  Gene Expression        4A  \n",
       "37  Gene Expression        4B  \n",
       "38  Gene Expression        4B  \n",
       "39  Gene Expression        5A  \n",
       "40  Gene Expression        5A  \n",
       "41  Gene Expression        5B  \n",
       "42  Gene Expression        5B  \n",
       "43  Gene Expression        5B  \n",
       "44  Gene Expression        6A  \n",
       "45  Gene Expression        6A  \n",
       "46  Gene Expression        6A  \n",
       "47  Gene Expression        6B  \n",
       "48  Gene Expression        6B  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gex_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group 10A:\n",
      "Group 10B:\n",
      "Group 11A:\n",
      "Group 11B:\n",
      "Group 12A:\n",
      "Group 12B:\n",
      "Group 13A:\n",
      "Group 13B:\n",
      "Group 14A:\n",
      "Group 14B:\n",
      "Group 2A:\n",
      "Group 2B:\n",
      "Group 3A:\n",
      "Group 3B:\n",
      "Group 4A:\n",
      "Group 4B:\n",
      "Group 5A:\n",
      "Group 5B:\n",
      "Group 6A:\n",
      "Group 6B:\n"
     ]
    }
   ],
   "source": [
    "df = pd.concat([gex_1, gbc_1, gex_2, gbc_2])\n",
    "\n",
    "# Group the DataFrame by 'GroupColumn'\n",
    "grouped = df.groupby('sample_id')\n",
    "\n",
    "# Export each group to separate DataFrames\n",
    "for group_name, group_data in grouped:\n",
    "    group_df = pd.DataFrame(group_data)\n",
    "    # Here, 'group_name' is the value in the 'GroupColumn' that represents the group.\n",
    "    # You can save/export 'group_df' to a file or store it in a data structure as needed.\n",
    "    print(f\"Group {group_name}:\")\n",
    "    group_df[['fastqs', 'sample', 'library_type']].to_csv('feature_extract_lib_{0}.csv'.format(group_name), index = False)\n",
    "    #print(group_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_extract_lib_10A.csv  feature_extract_lib_2A.csv\r\n",
      "feature_extract_lib_10B.csv  feature_extract_lib_2B.csv\r\n",
      "feature_extract_lib_11A.csv  feature_extract_lib_3A.csv\r\n",
      "feature_extract_lib_11B.csv  feature_extract_lib_3B.csv\r\n",
      "feature_extract_lib_12A.csv  feature_extract_lib_4A.csv\r\n",
      "feature_extract_lib_12B.csv  feature_extract_lib_4B.csv\r\n",
      "feature_extract_lib_13A.csv  feature_extract_lib_5A.csv\r\n",
      "feature_extract_lib_13B.csv  feature_extract_lib_5B.csv\r\n",
      "feature_extract_lib_14A.csv  feature_extract_lib_6A.csv\r\n",
      "feature_extract_lib_14B.csv  feature_extract_lib_6B.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastqs</th>\n",
       "      <th>sample</th>\n",
       "      <th>library_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>10A_GEX_IGO_15055_11</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055/F...</td>\n",
       "      <td>10A_guides_IGO_15055_31</td>\n",
       "      <td>CRISPR Guide Capture</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055_B...</td>\n",
       "      <td>10A_GEX_IGO_15055_B_11</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055_B...</td>\n",
       "      <td>10A_GEX_IGO_15055_B_11</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>/data/norman/southark/datasets/Project_15055_B...</td>\n",
       "      <td>10A_guides_IGO_15055_B_31</td>\n",
       "      <td>CRISPR Guide Capture</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              fastqs  \\\n",
       "0  /data/norman/southark/datasets/Project_15055/F...   \n",
       "1  /data/norman/southark/datasets/Project_15055/F...   \n",
       "2  /data/norman/southark/datasets/Project_15055/F...   \n",
       "3  /data/norman/southark/datasets/Project_15055/F...   \n",
       "4  /data/norman/southark/datasets/Project_15055_B...   \n",
       "5  /data/norman/southark/datasets/Project_15055_B...   \n",
       "6  /data/norman/southark/datasets/Project_15055_B...   \n",
       "\n",
       "                      sample          library_type  \n",
       "0       10A_GEX_IGO_15055_11       Gene Expression  \n",
       "1       10A_GEX_IGO_15055_11       Gene Expression  \n",
       "2       10A_GEX_IGO_15055_11       Gene Expression  \n",
       "3    10A_guides_IGO_15055_31  CRISPR Guide Capture  \n",
       "4     10A_GEX_IGO_15055_B_11       Gene Expression  \n",
       "5     10A_GEX_IGO_15055_B_11       Gene Expression  \n",
       "6  10A_guides_IGO_15055_B_31  CRISPR Guide Capture  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('feature_extract_lib_10A.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/norman/southark/datasets/Project_15055_B/FAUCI_0055/Sample_10A_GEX_IGO_15055_B_11'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('feature_extract_lib_10A.csv').fastqs[5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up bash script for bsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write script to generate bsub file with ref locations etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "requesting the following amount of memory: 256 GB\n",
      "Script file 'submit_jobs.sh' created. You can submit the jobs later using 'bash submit_jobs.sh'.\n"
     ]
    }
   ],
   "source": [
    "# Set the reference path\n",
    "ref_path = \"/your/reference/path\"\n",
    "\n",
    "# Path to the folder containing CSV files\n",
    "csv_folder = working_dir\n",
    "\n",
    "# Path to the control CSV file\n",
    "control_csv_path = \"/path/to/your/rpe1_tfs_crispra_w_controls.csv\"\n",
    "\n",
    "# Memory allocation, wall time, and number of cores settings\n",
    "# if total mem requested changes you should modify the run_cellranger_count.sh script\n",
    "memory_allocation = 4  # You can change this value\n",
    "wall_time = \"48:00\"  # You can change this value\n",
    "num_cores = 64  # You can change this value\n",
    "\n",
    "#calc total mem requested\n",
    "total_mem = memory_allocation*num_cores\n",
    "print(f\"requesting the following amount of memory: {total_mem} GB\")\n",
    "\n",
    "# Output script file\n",
    "script_file = \"submit_jobs.sh\"\n",
    "\n",
    "# Open the script file for writing\n",
    "with open(script_file, \"w\") as script_file_writer:\n",
    "    # Iterate through CSV files in the folder and write to the script file\n",
    "    for csv_file in os.listdir(csv_folder):\n",
    "        # Check if the file is a CSV file and follows the naming convention\n",
    "        if csv_file.endswith(\".csv\") and csv_file.startswith(\"feature_extract_lib_\"):\n",
    "            # Extract sample_id from the CSV file name\n",
    "            sample_id = csv_file.split(\"_\")[-1].split(\".\")[0]\n",
    "\n",
    "            # Construct the bsub command and write to the script file\n",
    "            bsub_command = f\"bsub -o {sample_id}.o -e {sample_id}.e -n {str(num_cores)} -R 'rusage[mem={str(memory_allocation)}]' -W {wall_time} -cwd {working_dir} /usr/bin/bash run_cellranger_count.sh {sample_id} {os.path.join(csv_folder, csv_file)} {ref_path} {control_csv_path}\\n\"\n",
    "            script_file_writer.write(bsub_command)\n",
    "\n",
    "# Inform the user\n",
    "print(f\"Script file '{script_file}' created. You can submit the jobs later using 'bash {script_file}'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#creation/modification of run_cellranger_count.sh script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Script file 'run_cellranger_count.sh' created.\n"
     ]
    }
   ],
   "source": [
    "script_content = \"\"\"#!/usr/bin/bash\n",
    "\n",
    "# Usage: run_cellranger_count.sh NAME CSV_FILE GENOME GUIDE_CSV LOCALCORES LOCALMEM\n",
    "\n",
    "NAME=$1\n",
    "CSV_FILE=$2\n",
    "GENOME=$3\n",
    "GUIDE_CSV=$4\n",
    "\n",
    "echo \"Name: $NAME\"\n",
    "echo \"CSV File: $CSV_FILE\"\n",
    "echo \"Genome: $GENOME\"\n",
    "echo \"Guide CSV: $GUIDE_CSV\"\n",
    "echo \"Local Cores: $LOCALCORES\"\n",
    "echo \"Local Memory: $LOCALMEM\"\n",
    "\n",
    "export PATH=/data/norman/southark/software/cellranger-7.1.0:$PATH\n",
    "\n",
    "# Uncomment the following line if you want to perform a sitecheck\n",
    "# cellranger sitecheck > sitecheck.txt\n",
    "\n",
    "cellranger count --id=$NAME --libraries=$CSV_FILE --transcriptome=$GENOME --feature-ref=$GUIDE_CSV --localcores={0} --localmem={1} --include-introns=false\n",
    "\"\"\".format(num_cores, int(total_mem*0.9)) #cellranger should only have 90% of total available\n",
    "\n",
    "# Specify the file path\n",
    "script_file_path = \"run_cellranger_count.sh\"\n",
    "\n",
    "# Write the content to the script file\n",
    "with open(script_file_path, \"w\") as script_file:\n",
    "    script_file.write(script_content)\n",
    "\n",
    "# Display a message indicating that the file has been created\n",
    "print(f\"Script file '{script_file_path}' created.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Job submission instructions\n",
    "\n",
    "log on to lilac and submit jobs with the following code snippets\n",
    "\n",
    "```\n",
    "#submit jobs with this:\n",
    "bsub -pack /working_dir/submit_cellranger_jobs.sh\n",
    "\n",
    "#live updates of submission status:\n",
    "watch -n0.01 bjobs\n",
    "\n",
    "#get list of running/submited jobss\n",
    "bjobs\n",
    "\n",
    "#kill currently running jobs (if something is not right)\n",
    "bjobs -o \"jobid stat\" | awk '{print $1}' | tail -n +2 | xargs bkill\n",
    "\n",
    "#info mem/usage of current job\n",
    "bjobs -l <job_id>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-analysis37]",
   "language": "python",
   "name": "conda-env-.conda-analysis37-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
